File: test_mavros/src/setpoint_raw_publisher.cpp
#include <ros/ros.h>
#include <mavros_msgs/PositionTarget.h>

int main(int argc, char** argv) {
    ros::init(argc, argv, "setpoint_raw_publisher");
    ros::NodeHandle nh;
    
    ros::Publisher setpoint_pub = nh.advertise<mavros_msgs::PositionTarget>(
        "/mavros/setpoint_raw/local", 10);

    ros::Rate rate(20.0); // Publishing rate of 20 Hz
    
    ROS_INFO("Setpoint Raw publisher node started. Publishing commands...");

    while (ros::ok()) {
        mavros_msgs::PositionTarget target_msg;
        target_msg.header.stamp = ros::Time::now();
        target_msg.header.frame_id = "map";

        target_msg.coordinate_frame = mavros_msgs::PositionTarget::FRAME_LOCAL_NED;

        // Enable only position control (ignore velocity and acceleration)
        target_msg.type_mask =
            mavros_msgs::PositionTarget::IGNORE_VX |
            mavros_msgs::PositionTarget::IGNORE_VY |
            mavros_msgs::PositionTarget::IGNORE_VZ |
            mavros_msgs::PositionTarget::IGNORE_AFX |
            mavros_msgs::PositionTarget::IGNORE_AFY |
            mavros_msgs::PositionTarget::IGNORE_AFZ |
            mavros_msgs::PositionTarget::IGNORE_YAW_RATE;

        // Set desired position (x, y, z) in meters
        target_msg.position.x = 2.0;
        target_msg.position.y = 5.0;
        target_msg.position.z = 3.5;

        // Set desired yaw in radians
        target_msg.yaw = 1.57; // 0 degrees
        
        ROS_INFO("Publishing Position: [x: %.2f, y: %.2f, z: %.2f] | Yaw: [yaw: %.2f]",
                 target_msg.position.x, target_msg.position.y, target_msg.position.z,
                 target_msg.yaw);

        setpoint_pub.publish(target_msg);

        ros::spinOnce();
        rate.sleep();
    }

    return 0;
}

File: test_mavros/scripts/setpoint_raw_publisher.py
#!/usr/bin/env python3

import rospy
from mavros_msgs.msg import PositionTarget

def publish_setpoint_raw():
    rospy.init_node('setpoint_raw_position_publisher', anonymous=True)
    raw_pub = rospy.Publisher('/mavros/setpoint_raw/local', PositionTarget, queue_size=10)

    rate = rospy.Rate(10)  # 10 Hz

    rospy.loginfo("Publishing position setpoints to /mavros/setpoint_raw/local")

    while not rospy.is_shutdown():
        raw_msg = PositionTarget()
        raw_msg.header.stamp = rospy.Time.now()
        raw_msg.coordinate_frame = PositionTarget.FRAME_LOCAL_NED

        # Type mask to ignore everything except position
        raw_msg.type_mask = (
            PositionTarget.IGNORE_VX | PositionTarget.IGNORE_VY | PositionTarget.IGNORE_VZ |  # Ignore velocity
            PositionTarget.IGNORE_AFX | PositionTarget.IGNORE_AFY | PositionTarget.IGNORE_AFZ |  # Ignore acceleration
            PositionTarget.IGNORE_YAW | PositionTarget.IGNORE_YAW_RATE  # Ignore yaw and yaw rate
        )

        # Set position (meters)
        raw_msg.position.x = 2.0
        raw_msg.position.y = 5.0
        raw_msg.position.z = 3.5

        raw_pub.publish(raw_msg)
        rospy.loginfo(f"Publishing -> Position: ({raw_msg.position.x}, {raw_msg.position.y}, {raw_msg.position.z})")

        rate.sleep()

if __name__ == '__main__':
    try:
        publish_setpoint_raw()
    except rospy.ROSInterruptException:
        rospy.logwarn("Setpoint Raw Position publisher interrupted.")

